Areas of D usage

wilzbach opened the pull request:
As the _first five minutes matter_, I thought it would be nice if we have an overview of areas in which D is used.
It is different to the great [Overview](http://dlang.org/overview.html) page as it tries for every area to inform about
-  applications written in D
-  key reasons/selling points

That being said, this document is a draft and your feedback & review is highlighy appreciated.
If I left out an area or you miss a great application - please let me know!

Also note that I propose to put this on dlang.org, because in contrast to the wiki we can go through the very helpful peer-review stage and have the official branding.

wilzbach commented:
Can someone have a look at the content?
I beliebe providing such an overview easily accessible from the front page is part of our mission to make D more mainstream!

schuetzm commented:
Agreed that this is important. I wouldn't put "Bare metal" at the top of the list, though. It's not D's strongest point currently. Also, "Compilers" are not really bare metal applications.

schuetzm commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,178 @@
+    Remedy Entertainment has successfully shipped the first $(HTTP www.quantumbreak.com/,
+    AAA game) to use D code for XBox One and Windows 10. $(BR)
+    For more details see Ethan Watson's
+    $(HTTP http://dconf.org/2016/talks/watson.html, presentation).
```
The `http://` prefix mustn't be used with `HTTP`.

schuetzm commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,178 @@
+    features like Fibers and compile-time function execution makes it well suited
+    for high-performance asynchronous I/O, concurrency and web applications like
+    $(HTTPS https://github.com/rejectedsoftware/vibe.d, vibe.d). $(BR)
+
```
ditto

schuetzm commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,178 @@
+
+    As an example, have a look at this talk about a
+    $(HTTPS https://www.youtube.com/watch?v=_gCFpHxAJRc, Mobile Gameserver Backend in D)
+)
```
ditto

schuetzm commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,178 @@
+    as they would have been written in C.
+    Additionally due to D's introspection at compile-time it enables
+    further optimizations for superior performance. See Andrei's talk
+    $(HTTPS https://www.youtube.com/watch?v=AxnotgLql0k, Fastware) on this topic.
```
ditto

schuetzm commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,178 @@
+    working with GPUs for general purpose computation can appear daunting and laborious.
+    Even seasoned users will admit that the performance improvements are often not worth the extra effort. $(BR)
+    Thanks to the power the power of D’s type system and static introspection
+    $(HTTPS http://dconf.org/2016/talks/colvin.html, GPU Programming) is
```
Don't use `https://`

schuetzm commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,178 @@
+    further optimizations for superior performance. See Andrei's talk
+    $(HTTPS https://www.youtube.com/watch?v=AxnotgLql0k, Fastware) on this topic.
+
+    D's standard library includes support for $(HTTPS $(ROOT_DIR)phobos/std_experimental_ndslice.html, multidimensional arrays)
```
This results in a broken link, see Vladimir's auto tester.

schuetzm commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,178 @@
+    D's multi-paradigm (imperative, structured, object oriented, generic, functional programming
+    purity, and even assembly) approach allows to teach in one language and gradually
+    explain new features without needing to switch to a different language.
+    D also has a memory-safe subset $(I $(WEB ./safed.html, SafeD)). It's safety and
```
This link is broken, too, see the auto tester.

wilzbach commented:
> Agreed that this is important. I wouldn't put "Bare metal" at the top of the list, though.

How would you structure it?

> It's not D's strongest point currently. Also, "Compilers" are not really bare metal applications.

Renamed the category to System programming, fixed all the links and went over the text again.

schuetzm commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,189 @@
+from various areas. $(BR)
+
+For an overview of special features of D, have a look at the
+$(WEB $(ROOT_DIR)overview.html, D Overview page).
```
This one is still broken, need to use `LINK2` as below...

schuetzm commented:
I'd say by "popularity" of the usage area. For example "Numerical computing / Data science" seems to be a hot topic, judging from the forum activity. On the other hand, "Game development" is sexy. But we probably shouldn't make a science out of it.

wilzbach commented:
> I'd say by "popularity" of the usage area. For example "Numerical computing / Data science" seems to be a hot topic, judging from the forum activity. On the other hand, "Game development" is sexy

I put academia on purpose on the bottom to avoid the impression it's an "academic language". Maybe we can create a new grouping?
We could split academia into research and teaching, but then teaching would be alone.
Has someone else a better idea?

> This one is still broken, need to use LINK2 as below...

Rebased & fixed.

jmh530 commented:
The spacing on GPU programming looks wonky due to it being justified.

Also, I wouldn't necessarily group Numerical Computing, GPU Computing, and Data Science in to Academia. Maybe split them in to their own category. I might also mention how chaining range operations is like dplyr or something like that in the data science section.

aG0aep6G commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,189 @@
+)
+
+This list highlights a selection of great applications written in D
+from various areas. $(BR)
```
I dislike these `$(BR)`s throughout. I'd rather see paragraphs wrapped in `$(P ...)`. In cases where that adds to much space, I'd say just join the lines, as there is no need for separation then.

aG0aep6G commented:
This should be linked from somewhere. Documentation→Articles? Resources? I see that overview.html is apparently only linked from the Learn tour. That's not exactly great.

wilzbach commented:
>The spacing on GPU programming looks wonky due to it being justified.

> Also, I wouldn't necessarily group Numerical Computing, GPU Computing, and Data Science in to Academia. Maybe split them in to their own category.

It's now named cutting-edge research.

> I might also mention how chaining range operations is like dplyr or something like that in the data science section.

On it - will come soon. Thanks for the idea.

> This should be linked from somewhere. 

I know added overview and this article to the bottom of dlang.org - I think it fits, because the heading is "Why D?". However the wording isn't the best yet.

> I see that overview.html is apparently only linked from the Learn tour. That's not exactly great.

I know (I made many PRs to the tour) - we are highly "underlinked"!

> Resources?

We have to be careful the current number of menu items is already too much.
Can we somehow create a new group?

> Articles

Could be an idea, but probably not the first place I search for an overview.

> Documentation.

We should merge http://dlang.org/comparison.html (linked from documentation) with http://dlang.org/overview.html

wilzbach commented:
I had another look through the text and fixed a couple of points.

Btw that's how I propose to list the overview on the front page:

![image](https://cloud.githubusercontent.com/assets/4370550/15806528/5b088370-2b45-11e6-8ddd-02f91cb9336e.png)

How should we proceed? More nitpicks?

wilzbach commented:
Hey I went through the text again (minor revisions), but I added fancy icons - they are not optimal yet. The usual CSS hacks (responsive, nice alignment) are missing, however I think you could already check whether some icons don't fit

1) overview

![image](https://cloud.githubusercontent.com/assets/4370550/15904153/2818b026-2db0-11e6-8bc8-d4882f06ceb1.png)

2) sections

![image](https://cloud.githubusercontent.com/assets/4370550/15904190/4f283eac-2db0-11e6-812d-1c3646b4294e.png)

See [all](http://dtest.thecybershadow.net/artifact/website-fd5e6b914ec31f9e01b72737bc74f26d15dd8786-fa58e6418d1c0a99720e7479fc8c8d17/web/areas-of-d-usage.html)

CyberShadow commented:
This looks really cool but I think should be approved by W/A. Normally this would probably belong on the wiki, but I really like how it's laid out.

* Some weird wrapping here, maybe just use a constant margin for the whole section:

![](https://dump.thecybershadow.net/ab1b9c1d8a391c503fe589ddf79221a4/screenshot.png)

* The icon currently used for "Industry" seems more associated with academy, I think.

JackStouffer commented on areas-of-d-usage.dd:
```diff
@@ -0,0 +1,278 @@
+$(SECTION3 $(LNAME2 web, Web applications), $(ARTICLE_FA_ICON cloud)
+
+    D's ability to combine low and high-level code in one language, asynchronous
+    features like Fibers and compile-time function execution makes it well suited
```
I would say "Fibers (a.k.a coroutines)"

ZombineDev replied:
Well according to [N4024](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4024.pdf) we have both - `core.thread.Fiber` looks like `boost::coroutine` and **vibe.d**'s use of fibers is more like `boost::fiber`.

So I would say "stack-full coroutines and fibers".

wilzbach commented:
Last week I tried to make the icons more responsive. It's not perfect, but should look a lot better than before.
Moreover I add two nitpick runs over the text - can someone take 5-10 minutes to give the text a final nitpick round, s.t. we can ship an initial version? :)

JackStouffer commented:
I think using [this](http://fontawesome.io/icon/industry/) for the industry icon would make more sense.

JackStouffer commented:
Why is the industry icon on the other side of the page?
<img width="1254" alt="screen shot 2016-06-15 at 9 29 06 am" src="https://cloud.githubusercontent.com/assets/680068/16081498/a3d58236-32db-11e6-8bfb-7c9f09e3b381.png">

JackStouffer commented:
I would remove the word "perfect" from the industry section. Your aim as a programmer is never to make perfect code as you know that's impossible.

wilzbach commented:
> I think using this for the industry icon would make more sense.

It isn't part of FontAwesome 4.2  (see also #1371) - I guess to whether icon we decide there, I will just use the same here.

> Why is the industry icon on the other side of the page?

The idea was to have indent the subcategories, so I put for the categories the icons on the right side.
Is this too weird? Better ideas?

JackStouffer commented:
"few scientific programmers care about" sounds negative. "few scientific programs need to worry about" sounds better IMO.

Also I would mention parallelism somewhere in that section.

JackStouffer commented:
"Last but least" -> "Last but not least"

JackStouffer commented:
I would link to the orgs using D page somewhere in the industry section

JackStouffer commented:
> Is this too weird? Better ideas?

Yeah, it's kind of jarring. This effect is already achieved via the typography of the section headers.

wilzbach commented:
> Yeah, it's kind of jarring. This effect is already achieved via the typography of the section headers.

Alrighty :)

> Also I would mention parallelism somewhere in that section.

I added:

"    allows to easily parallelize your algorithm and pipelines,"

> I would link to the orgs using D page somewhere in the industry section

I added:

"D has been used in many, diverse domains. A short selection will be presented below.
 For a full overview you can browse the list of reported $(LINK2 $(ROOT_DIR)orgs-using-d.html, organizations which use the D Language).

Thanks @JackStouffer :)
Do we need more eyes or is this ready for the first version?

andralex commented:
I guess we're ready. Should we take away the funny but aged D mascot?

JackStouffer commented:
Please squash commits

wilzbach commented:
> I guess we're ready. Should we take away the funny but aged D mascot?

There is a related discussion for the DLang Tour:

https://github.com/stonemaster/dlang-tour/pull/248

> Please squash commits

With pleasure :)

andralex commented:
Engage!
